module/zstd/zfs_zstd.c

```diff
@@ -202,6 +202,25 @@ static struct zstd_fallback_mem zstd_dctx_fallback;
 static struct zstd_pool *zstd_mempool_cctx;
 static struct zstd_pool *zstd_mempool_dctx;
 
+/*
+ * The library zstd code expects these if ADDRESS_SANITIZER gets defined,
+ * and while ASAN does this, KASAN defines that and does not. So to avoid
+ * changing the external code, we do this.
+ */
+#if defined(__has_feature)
+#if __has_feature(address_sanitizer)
+#define	ADDRESS_SANITIZER 1
+#endif
+#elif defined(__SANITIZE_ADDRESS__)
+#define	ADDRESS_SANITIZER 1
+#endif
+#if defined(_KERNEL) && defined(ADDRESS_SANITIZER)
+void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
+void __asan_poison_memory_region(void const volatile *addr, size_t size);
+void __asan_unpoison_memory_region(void const volatile *addr, size_t size) {};
+void __asan_poison_memory_region(void const volatile *addr, size_t size) {};
```

Comment:
Is there a reason why we couldn't just define the function here?  i.e. just:

```suggestion
#if defined(_KERNEL) && defined(ADDRESS_SANITIZER)
void __asan_unpoison_memory_region(void const volatile *addr, size_t size) {};
void __asan_poison_memory_region(void const volatile *addr, size_t size) {};
#endif
```

Reply:
`/home/rich/zfs_vanilla/module/zstd/zfs_zstd.c:221:6: error: no previous prototype for ‘__asan_poison_memory_region’ [-Werror=missing-prototypes]` would be why.

Reply:
Ahh, yes I'd forgotten we added that check.  Can they be declared statically then?  (Not that it's a big deal either way)

```
static void __asan_unpoison_memory_region(void const volatile *addr, size_t size) {};
```

Reply:
Not unless I'm very bad at it:
`/home/rich/zfs_vanilla/module/zstd/zfs_zstd.c:221:13: error: ‘__asan_poison_memory_region’ defined but not used [-Werror=unused-function]
`

Reply:
We could probably move the forward declarations in to `include/sys/zstd/zstd.h`, although I see why you wanted to keep all of this in one spot.  This is a big odd but I'm good with it.  Thanks for humoring me.